onnxruntime/core/optimizer/gemm_transpose_fusion.cc

```diff
@@ -0,0 +1,116 @@
+  }
+  // check if input B is a Transpose
+  if (B_node_ptr != nullptr && B_node_ptr->OpType() == "Transpose") {
+    Node& B_node = *graph.GetNode(B_node_ptr->Index());
+    transB = !transB;
+    nodes_to_remove.push_back(B_node);
+    new_gemm_input_defs[1] = B_node.MutableInputDefs()[0];
+  }
+
+  nodes_to_remove.push_back(gemm_node);
+
+  // check if output node is Transpose
+  if (output_node_ptr != gemm_node.OutputNodesEnd() && 
+      gemm_node.InputDefs().size() <= 2 && // C is missing
+      output_node_ptr->OpType() == "Transpose") {
+    Node& output_node = *graph.GetNode(output_node_ptr->Index());
+    // (AB)' = B'A' : reverse the inputs
+    std::reverse(new_gemm_input_defs.begin(), new_gemm_input_defs.end());
+    transA = !transB;
+    transB = !transA;
+
+    nodes_to_remove.push_back(output_node);
+  }
+
+  Node& new_gemm_node = graph.AddNode(graph.GenerateNodeName(gemm_node.Name() + "_transformed"),
+                                      gemm_node.OpType(),
+                                      "Fused Gemm with Transpose",
+                                      new_gemm_input_defs,
+                                      {},
+                                      {},
+                                      gemm_node.Domain());
+  new_gemm_node.AddAttribute("transA", static_cast<int64_t>(transA));
+  new_gemm_node.AddAttribute("transB", static_cast<int64_t>(transB));
+  new_gemm_node.AddAttribute("alpha", gemm_node.GetAttributes().at("alpha").f());
+  new_gemm_node.AddAttribute("beta", gemm_node.GetAttributes().at("beta").f());
+
+  graph_utils::FinalizeNodeFusion(graph, nodes_to_remove, new_gemm_node);
+
+  modified = RewriteRuleEffect::kModifiedRestOfGraph;
+
+  return Status::OK();
+}
+
+bool GemmTransposeFusion::SatisfyCondition(const Graph& graph, const Node& node, const logging::Logger&) const {
+  if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "Gemm", {1, 6, 7, 9, 11, 13}) ||
+      node.GetOutputEdgesCount() > 1) {
+    return false;
+  }
+
+  // Fusion can be applied if there is a transpose at either of the inputs
+  for (auto node_it = node.InputNodesBegin(); node_it != node.InputNodesEnd(); ++node_it) {
+    if (graph_utils::IsSupportedOptypeVersionAndDomain(*node_it, "Transpose", {1, 13}) &&
+        node_it->GetOutputEdgesCount() == 1 &&
+        graph.GetNodeOutputsInGraphOutputs(*node_it).empty() &&
+        // Make sure the two nodes do not span execution providers.
+        node_it->GetExecutionProviderType() == node.GetExecutionProviderType()) {
+      return true;
+    }
+  }
+
+  // Fusion can be applied if there is a Transpose at the output of Gemm
+  // by the rule (AB)' = B'A' provided that C is missing
+  // Supported for Opset >=11 as earlier opsets have C as a required input
+  if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "Gemm", {11, 13}) ||
+      !graph.GetNodeOutputsInGraphOutputs(node).empty() ||
+      // verify that C is missing
+      node.InputDefs().size() > 2) {
```

Comment:
Optional: Does it make sense to also check for the case where C is present but zero? I'm not sure if this is a common case (after the recent PR to take advantage of the optional input in grad builders). Also arguably it should be handled by a separate transformation that runs before this one.

Reply:
we can add a separate transform that removes the C if it is 0s, which should be more effective as it would apply to all Gemms independent of Transpose patterns